lib/ipam/ipam.go

```diff
@@ -1383,6 +1343,91 @@ func (c ipamClient) convertBackendToIPAMConfig(cfg *model.IPAMConfig) *IPAMConfi
 	}
 }
 
+// ensureConsistentAffinity retrieves the pool and node for the given block and determines
+// if the pool still selects node. If it no longer matches, it will release the block
+// affinity for that node.
+// Returns a bool indicating if the block affinity was released.
+func (c ipamClient) ensureConsistentAffinity(ctx context.Context, b *model.AllocationBlock) (bool, error) {
+
+	// Retrieve node for this allocation. We do this so we can clean up affinity for blocks
+	// which should no longer be affine to this host.
+	host := getHostAffinity(b)
+	logCtx := log.WithFields(log.Fields{"cidr": b.CIDR, "host": host})
+
+	// If no hostname is found on the block affinity,
+	// there is no need to do an ip pool node selection check.
+	if host == "" {
+		logCtx.Debug("Block already has no affinity")
+		return false, nil
+	}
+
+	// If the IP pool which owns this block no longer selects this node,
+	// we should release the block's affinity to this node so it can be
+	// used elsewhere.
+	logCtx.Debugf("Looking up node labels for host affinity")
+	node, err := c.client.Get(ctx, model.ResourceKey{Kind: v3.KindNode, Name: host}, "")
+	if err != nil {
+		if _, ok := err.(cerrors.ErrorResourceDoesNotExist); !ok {
+			logCtx.WithError(err).WithField("node", host).Error("Failed to get node for host")
+			return false, err
+		}
+		logCtx.Info("Node doesn't exist, releasing IPs anyway")
+		return false, nil
+	}
+
+	// Make sure the returned value is a valid node.
+	v3n, ok := node.Value.(*v3.Node)
+	if !ok {
+		return false, fmt.Errorf("Datastore returned malformed node object")
+	}
+
+	// Fetch the pool for the given CIDR and check if it selects the node.
+	if pool := c.blockReaderWriter.getPoolForIP(cnet.IP{b.CIDR.IPNet.IP}); pool == nil {
+		logCtx.Debug("No pools own this block")
+		return false, nil
+	} else if sel, err := pool.SelectsNode(*v3n); err != nil {
+		logCtx.WithField("selector", pool.Spec.NodeSelector).WithError(err).Error("Failed to determine node selection")
+		return false, err
+	} else if sel {
+		logCtx.Debug("Pool selects node, no change")
+		return false, nil
+	} else {
+		logCtx.WithField("selector", pool.Spec.NodeSelector).Debug("Pool no longer selects node, releasing block affinity")
```

Comment:
In this case I think moving this log statement outside of the else and putting it right above the call the `c.releaseBlockAffinity` would reduce the number of lines of code, and move the log statement closer to the thing it is actually describing. We can do that here since all of the other if / else if blocks have returns in them.

Reply:
bah, again it's cuz we're using the pool here. For some reason I can't see that ever.